lib/src/git.rs

```diff
@@ -110,6 +115,37 @@ fn to_remote_branch<'a>(parsed_ref: &'a RefName, remote_name: &str) -> Option<&'
     }
 }
 
+// there are two options:
+//
+// a bare git repo, which has a parent named .jj that sits on the workspace root
+// or
+// a colocated .git dir, which is already on the workspace root
+fn work_tree_from_git_dir(git_dir: &Path) -> Result<PathBuf, String> {
+    if git_dir.file_name() == Some(OsStr::new(".git")) {
+        git_dir
+            .parent()
+            .map(|x| x.to_path_buf())
+            .ok_or(format!("git repo had no parent: {}", git_dir.display()))
+    } else if git_dir.file_name() == Some(OsStr::new("git")) {
+        let mut it = git_dir.ancestors();
+        for path in it.by_ref() {
+            if path.file_name() == Some(OsStr::new(".jj")) {
+                break;
+            }
+        }
+
+        it.next().map(|x| x.to_path_buf()).ok_or(format!(
+            "could not find .jj dir in git dir path: {}",
+            git_dir.display()
+        ))
```

Comment:
suggestion: Perhaps you want something like `it.skip_while(/* isn't .jj */).next()`? (Assuming that this function makes it into the final version.)